[SPARK-27180][BUILD][YARN] Fix testing issues with yarn module in Hadoop-3

wangyum opened the pull request:
## What changes were proposed in this pull request?

Fix testing issues with `yarn` module in Hadoop-3:

1. Upgrade jersey-1 to `1.19` to fix ```Cause: java.lang.NoClassDefFoundError: com/sun/jersey/spi/container/servlet/ServletContainer```.
2. Copy `ServerSocketUtil` from hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/net/ServerSocketUtil.java to fix ```java.lang.NoClassDefFoundError: org/apache/hadoop/net/ServerSocketUtil```.
3. Adapte `SessionHandler` from jetty-9.3.25.v20180904/jetty-server/src/main/java/org/eclipse/jetty/server/session/SessionHandler.java  to fix ```java.lang.NoSuchMethodError: org.eclipse.jetty.server.session.SessionHandler.getSessionManager()Lorg/eclipse/jetty/server/SessionManager```.

## How was this patch tested?

manual tests:
```shell
build/sbt yarn/test -Pyarn
build/sbt yarn/test -Phadoop-3.2 -Pyarn

build/mvn -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnClusterSuite -pl resource-managers/yarn test -Pyarn
build/mvn -Dtest=none -DwildcardSuites=org.apache.spark.deploy.yarn.YarnClusterSuite -pl resource-managers/yarn test -Pyarn -Phadoop-3.2
```

wangyum committed:
Fix hadoop-3 yarn test

wangyum commented on resource-managers/yarn/pom.xml:
```diff
@@ -29,7 +29,7 @@
   <properties>
     <sbt.project.name>yarn</sbt.project.name>
-    <jersey-1.version>1.9</jersey-1.version>
+    <jersey-1.version>1.19</jersey-1.version>
```
Upgrade jersey to 1.19. otherwise:
```
[info]   Cause: java.lang.NoClassDefFoundError: com/sun/jersey/spi/container/servlet/ServletContainer
[info]   at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager.startWepApp(ResourceManager.java:1103)
[info]   at org.apache.hadoop.yarn.server.resourcemanager.ResourceManager.serviceStart(ResourceManager.java:1271)
[info]   at org.apache.hadoop.service.AbstractService.start(AbstractService.java:194)
```

wangyum commented on resource-managers/yarn/src/test/java/org/apache/hadoop/net/ServerSocketUtil.java:
```diff
@@ -0,0 +1,132 @@
+/*
```
Add this class. otherwise:
```
[info] YarnClusterSuite:
[info] org.apache.spark.deploy.yarn.YarnClusterSuite *** ABORTED *** (33 milliseconds)
[info]   java.lang.NoClassDefFoundError: org/apache/hadoop/net/ServerSocketUtil
[info]   at org.apache.hadoop.yarn.server.MiniYARNCluster.serviceInit(MiniYARNCluster.java:260)
```
I try to do it by maven, but failed. It seems that [`sbt-pom-reader` does not support `test-jar`](https://github.com/sbt/sbt-pom-reader/pull/#14) very well.:

```xml
    <dependency>
      <groupId>org.apache.hadoop</groupId>
      <artifactId>hadoop-common</artifactId>
      <version>${hadoop.version}</version>
      <type>test-jar</type>
      <scope>test</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```

vanzin replied:
We do have other test-jar imports; some use `type` and some `classifier`. Have you tried the latter? e.g. we have

```
    <dependency>
      <groupId>org.apache.hadoop</groupId>
      <artifactId>hadoop-yarn-server-tests</artifactId>
      <classifier>tests</classifier>
      <scope>test</scope>
    </dependency>
```

wangyum replied:
Yes, I tried it, this can be successful on hadoop-3.2, but throws a compilation exception on hadoop-2.7(It seems exclude `hadoop-common`).
```java
[error] /Users/yumwang/SPARK-20845/spark/resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/ApplicationMasterSuite.scala:34: Class org.apache.hadoop.conf.Configuration not found - continuing with a stub.
[error]     val yarnConf = new YarnConfiguration()
[error] 
```

wangyum commented on pom.xml:
```diff
@@ -2651,6 +2651,7 @@
         <hadoop.version>3.1.0</hadoop.version>
         <curator.version>2.12.0</curator.version>
         <zookeeper.version>3.4.9</zookeeper.version>
+        <jetty.version>9.3.24.v20180605</jetty.version>
```
Downgrade it, otherwise:
```
[info] YarnClusterSuite:
[info] org.apache.spark.deploy.yarn.YarnClusterSuite *** ABORTED *** (177 milliseconds)
[info]   org.apache.hadoop.yarn.exceptions.YarnRuntimeException: java.lang.NoSuchMethodError: org.eclipse.jetty.server.session.SessionHandler.getSessionManager()Lorg/eclipse/jetty/server/SessionManager;
```
Furthermore, We have some discuss about this change:
https://issues.apache.org/jira/browse/HADOOP-16152

srowen replied:
Hm, I'm afraid this might conflict with Java 11 compatibility, as this was updated for that reason: https://github.com/apache/spark/pull/22993
https://www.eclipse.org/lists/jetty-announce/msg00124.html

I wonder if this can be worked around on the Spark side, or whether it is indeed possible for Hadoop 3 to update?

vanzin replied:
We shouldn't change Spark's version of Jetty because a YARN test is failing. A Spark application will not hit the same code paths as the YARN tests (which run a YARN server).

Maybe change the version of jetty in the YARN module when testing with hadoop-3, if there's no other way to work around this.

wangyum replied:
I tried several ways but they all failed:
1. Use `hadoop-client-minicluster` instead of `hadoop-yarn-server-tests`: https://github.com/wangyum/spark-hadoop-client-minicluster
2. Change the version of jetty in the YARN module when testing with hadoop-3
3. Upgrade Eclipse Jetty version to 9.4.x for Hadoop: https://issues.apache.org/jira/browse/HADOOP-16152

srowen replied:
The ideal outcome is updating Hadoop, as otherwise it seems like Hadoop 3 and Java 11 support are in conflict. Then again, if Hadoop 3.x isn't quite going to work with Java 11 for other reasons, we have larger problems. Hadoop 2.x is currently working OK with Java 11 tests (it's Hive that's the issue) so I'm kind of surprised.

How did you try to override the Jetty version in the YARN module? I'd expect that's entirely possible

wangyum replied:
I try to  override the Jetty version in the YARN module by: https://github.com/apache/spark/pull/24115/commits/29e583f59685f1343f618269b22c802f7ed553df.
But ```+-org.eclipse.jetty:jetty-servlet:9.3.24.v20180605 (evicted by: 9.4.12.v20180830)```.

I found that adapted `SessionHandler` from jetty-9.3.25.v20180904 can test the YARN module in Hadoop 2 and Hadoop 3. I know this is not a good way. But it seems that this is the only way at the moment.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8972/
Test PASSed.

SparkQA commented:
**[Test build #103561 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103561/testReport)** for PR 24115 at commit [`757a09a`](https://github.com/apache/spark/commit/757a09abd7ece4100ae5b1225f86a118f84a178d).

SparkQA commented:
**[Test build #103561 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103561/testReport)** for PR 24115 at commit [`757a09a`](https://github.com/apache/spark/commit/757a09abd7ece4100ae5b1225f86a118f84a178d).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class ServerSocketUtil `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103561/
Test FAILed.

wangyum committed:
Merge remote-tracking branch 'upstream/master' into hadoop3-yarn

# Conflicts:
#	pom.xml

wangyum committed:
Add test jetty version

wangyum committed:
Add test jetty version, bit it does not work:
[0m[[0minfo[0m] [0m  | | | | +-org.eclipse.jetty:jetty-servlet:9.3.24.v20180605 (evicted by: 9.4.12.v20180830)[0m

wangyum committed:
Adapted SessionHandler from jetty-9.3.25.v20180904

SparkQA commented:
**[Test build #104104 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104104/testReport)** for PR 24115 at commit [`3c9bb7b`](https://github.com/apache/spark/commit/3c9bb7b4eb9828a0823b0b68080b8accca4a18b4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9450/
Test PASSed.

SparkQA commented:
**[Test build #104104 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104104/testReport)** for PR 24115 at commit [`3c9bb7b`](https://github.com/apache/spark/commit/3c9bb7b4eb9828a0823b0b68080b8accca4a18b4).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class SessionHandler extends ScopedHandler `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104104/
Test FAILed.

wangyum commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9454/
Test PASSed.

SparkQA commented:
**[Test build #104108 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104108/testReport)** for PR 24115 at commit [`3c9bb7b`](https://github.com/apache/spark/commit/3c9bb7b4eb9828a0823b0b68080b8accca4a18b4).

SparkQA commented:
**[Test build #104108 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104108/testReport)** for PR 24115 at commit [`3c9bb7b`](https://github.com/apache/spark/commit/3c9bb7b4eb9828a0823b0b68080b8accca4a18b4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class SessionHandler extends ScopedHandler `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104108/
Test PASSed.

srowen commented on resource-managers/yarn/src/test/java/org/eclipse/jetty/server/SessionManager.java:
```diff
@@ -0,0 +1,289 @@
+import org.eclipse.jetty.util.component.LifeCycle;
+
+/**
+ * Adapted from https://github.com/eclipse/jetty.project/blob/jetty-9.3.25.v20180904/
```
OK yeah, seems like the least-bad option to get this working.

felixcheung replied:
I think if the original file has a license, we can't replace or change it to Apache v2 at the topic.

FYI, jetty eclipse license here https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/LICENSE

wangyum replied:
It will fail when running RAT checks:
```
========================================================================
Running Apache RAT checks
========================================================================
Attempting to fetch rat
Could not find Apache license headers in the following files:
 !????? /home/jenkins/workspace/SparkPullRequestBuilder/resource-managers/yarn/src/test/java/org/eclipse/jetty/server/SessionManager.java
 !????? /home/jenkins/workspace/SparkPullRequestBuilder/resource-managers/yarn/src/test/java/org/eclipse/jetty/server/session/SessionHandler.java
[error] running /home/jenkins/workspace/SparkPullRequestBuilder/dev/check-license ; received return code 1
```

srowen replied:
If these are a straight copy, then it needs the original source's copyright to be retained as-is, yes, and we can't add an ASF header. You can exclude this file from RAT in `dev/.rat-excludes` (You are welcome to update it to 0.13 too while you're at it.)

Handling of EPL v1.0 is a little more nuanced:
https://apache.org/legal/resolved.html#weak-copyleft-licenses

The notes don't quite cover this case, of test-only code. The logic is that we want to avoid users creating a derivative work of this code. I don't see any case where a downstream user would modify this or include it in a software product; I can imagine someone bundling the Spark test code as-is, but not depending on it.

I'll ask legal if there's more to it.

srowen replied:
Actually, it's dual licensed under the Apache License 2.0 and EPL v1.0. We need to keep the original header and exclude the RAT check in both files, but there's no other issue here.

wangyum committed:
Update License

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9469/
Test PASSed.

SparkQA commented:
**[Test build #104123 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104123/testReport)** for PR 24115 at commit [`a33fcef`](https://github.com/apache/spark/commit/a33fcef3b04dd21a32269c57aa8a5d3eb5124231).

SparkQA commented:
**[Test build #104123 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104123/testReport)** for PR 24115 at commit [`a33fcef`](https://github.com/apache/spark/commit/a33fcef3b04dd21a32269c57aa8a5d3eb5124231).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104123/
Test FAILed.

wangyum committed:
revert to apache License

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9470/
Test PASSed.

SparkQA commented:
**[Test build #104124 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104124/testReport)** for PR 24115 at commit [`6a00f19`](https://github.com/apache/spark/commit/6a00f193a295e4bb928ce6cf034ddd9af9ceddf8).

SparkQA commented:
**[Test build #104124 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104124/testReport)** for PR 24115 at commit [`6a00f19`](https://github.com/apache/spark/commit/6a00f193a295e4bb928ce6cf034ddd9af9ceddf8).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104124/
Test FAILed.

wangyum commented:
retest this please

SparkQA commented:
**[Test build #104136 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104136/testReport)** for PR 24115 at commit [`6a00f19`](https://github.com/apache/spark/commit/6a00f193a295e4bb928ce6cf034ddd9af9ceddf8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9479/
Test PASSed.

SparkQA commented:
**[Test build #104136 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104136/testReport)** for PR 24115 at commit [`6a00f19`](https://github.com/apache/spark/commit/6a00f193a295e4bb928ce6cf034ddd9af9ceddf8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104136/
Test PASSed.

wangyum committed:
Fix LICENSE

SparkQA commented:
**[Test build #104149 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104149/testReport)** for PR 24115 at commit [`49f0d8a`](https://github.com/apache/spark/commit/49f0d8a49c1c75775d22e7c4117e854a9facbfa6).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9492/
Test PASSed.

srowen commented on dev/.rat-excludes:
```diff
@@ -115,3 +115,5 @@ structured-streaming/*
 kafka-source-initial-offset-version-2.1.0.bin
 kafka-source-initial-offset-future-version.bin
 vote.tmpl
+SessionManager.java
```
I guess we can't narrow this to the particular file in its particular directory? it's probably fine if not

wangyum replied:
Yes.  Please see [RAT-240](https://issues.apache.org/jira/browse/RAT-240) CLI: RAT excludes file does not allow to set a full path

srowen reviewed (approved):
Looks OK if that's what we have to do here.

SparkQA commented:
**[Test build #104149 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104149/testReport)** for PR 24115 at commit [`49f0d8a`](https://github.com/apache/spark/commit/49f0d8a49c1c75775d22e7c4117e854a9facbfa6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104149/
Test PASSed.

srowen commented:
Merged to master
